controller/src/main/java/org/jboss/as/controller/capability/RuntimeCapability.java

```diff
@@ -562,13 +554,35 @@ public Builder<T> setServiceType(Class<?> type) {
          */
         public Builder<T> addRequirements(String... requirements) {
             assert requirements != null;
-            if (this.requirements == null) {
-                this.requirements = new HashSet<>(requirements.length);
+            if (requirements.length > 0) {
+                this.requirements = this.requirements.isEmpty() ? Set.of(requirements) : Stream.concat(this.requirements.stream(), Stream.of(requirements)).collect(Collectors.toUnmodifiableSet());
             }
-            Collections.addAll(this.requirements, requirements);
             return this;
         }
 
+        /**
+         * Adds the names of other capabilities that this capability requires. The requirement
+         * for these capabilities will automatically be registered when this capability is registered.
+         *
+         * @param requirements the capability names
+         * @return the builder
+         */
+        public Builder<T> addRequirements(NullaryServiceDescriptor<?>... requirements) {
+            assert requirements != null;
+            switch (requirements.length) {
+                case 0:
+                    return this;
+                case 1:
+                    return this.addRequirements(requirements[0].getName());
+                case 2:
+                    return this.addRequirements(requirements[0].getName(), requirements[1].getName());
+                default:
+                    Stream<String> mapped = Stream.of(requirements).map(NullaryServiceDescriptor::getName);
+                    this.requirements = (this.requirements.isEmpty() ? mapped : Stream.concat(this.requirements.stream(), mapped)).collect(Collectors.toUnmodifiableSet());
```

Comment:
> if this.requirements.isEmpty() is true, this.requirements should be assigned to an unmodificable Set, however Stream<String> mapped is not an unmodificable set at this point.

I'm not sure what you mean.  Both cases use the same collector that returns an unmodifiable set.

Reply:
> Also, out of curiosity, why not use the code for the default: case always and remove the need for a switch? I mean, why use case 1 and 2?

Because we want to avoid allocating a hash table for only 1 or 2 entries, if possible.

Reply:
@pferraro 

> I'm not sure what you mean. Both cases use the same collector that returns an unmodifiable set.

Forget it, I see it now, sorry for the noise about that.